### ISSUE TITLE
Allow GUI un/installs

### ISSUE THREAD

**Issue by vitorgalvao** [caskroom/homebrew-cask#24377]
Ever since the inception of Homebrew-Cask, we (mostly<sup id="a1">[1](#f1)</sup>) haven’t allowed casks to call GUIs when installing and uninstalling, in the name of automation.

However, in some cases it actually prevents desirable automation: not auto-opening [`installer manual:`](https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/installer.md#installer-manual) targets and not using tool-provided uninstallers come to mind.

I postulate most users would actually prefer these GUI actions to be called automatically in most cases, the most notable exception being when reinstalling a system (since a bunch of casks are installed).

It also seems more in line with the goal of #13201 of making our automation as close to the manual installation as possible.

---

Proposal:

Change policy and start allowing GUIs to be called, *but only allow GUIs as the last resort*.

When calling a GUI is necessary for an action, allow the user to pick if a GUI will be called or the action will be interrupted. Make allowing GUIs the default.

Implement `--no-gui` as a flag to be given at runtime or in (`HOMEBREW_CASK_OPTS`). It means “If this action needs a GUI, don’t run it, and instead output a warning on how to proceed manually” (i.e. the same instructions we have now).

Implement `installing_calls_gui` and `uninstalling_calls_gui`. Similarly to `auto_updates` and `caskroom_only`, this should only exist when the value is `true`. This serves a single purpose: when we’re calling `install`/`uninstall` and the corresponding `*_calls_gui` stanza is `true`, check is `--no-gui` is set and act accordingly. In terms of [stanza order](https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_language_reference/readme.md#stanza-order), right above `auto_updates` seems fine.

When opening GUIs, do so with `open -W`. This way the command will be blocked until it finishes, preventing a barrage of GUIs left open when installing multiple apps and allowing for `&&` and other subsequent commands.

Rename `installer manual:` to `installer gui:`.

Add `uninstall gui:`. It runs *in addition to* the other flags, and before them (otherwise it could very well be removed by them).

---

Checklist:

+ [ ] Add `installing_calls_gui` and `uninstalling_calls_gui` stanzas.
+ [ ] Add `--no-gui` flag.
+ [ ] Rename (and update implementation of) `installer manual:`.
+ [ ] Implement `uninstall gui:`.
+ [ ] Update documentation.
+ [ ] Add tests.
+ [ ] Add rubocop cops.

---

<sup id="f1">[1](#a1)</sup> Some have slipped through on occasion, and some are still present, even.

**Comment by reitermarkus** [comment]
Since `installer manual:` seems to be used exclusively for GUI installers anyway, so for uninstalling, I'd name it `uninstall(er?) manual:`.

We also don't need `*_calls_gui`, because with `manual:`, this will always be the case.

**Comment by vitorgalvao** [comment]
> We don't need `*_calls_gui`, because with `installer/uninstall gui:`, we already know this.

Good point, completely missed that, as the ideas to rename and add `gui:` came when I was already writing. Amended the post.

**Comment by vitorgalvao** [comment]
Pinging @caskroom/maintainers, but any user is welcome to chime in.

**Comment by mwean** [comment]
I think this is a great direction 👍

**Comment by Numbermaniac** [comment]
Perhaps a note should be added to the relevant brew cask info _cask_ if the installation requires a GUI?

**Comment by vitorgalvao** [comment]
@Numbermaniac Agreed. Updated the top post.

**Comment by jawshooah** [comment]
I'm fine with this as long as we aren't blocking on the GUI installer. Otherwise we'll break automated install scripts.

**Comment by vitorgalvao** [comment]
> Otherwise we'll break automated install scripts.

How so? That’s what `--no-gui` is for.

If you say we should `--yes-gui` instead, at least at first (giving a warning about the reversal), I can agree, but being completely non-blocking on GUI installs sounds like a bad idea to me, *especially* when installing multiple apps and on automated install scripts. If I have an automated install script, I don’t want a GUI to suddenly popup in front of me if I haven’t explicitly called it. Similarly, when installing multiple apps I don’t want a ton of GUIs to suddenly flood the screen. Also, making it non-blocking makes scripting less useful, since we can’t rely on the installation being finished before continuing. As an example, imagine an automated script that installs the cask and then reboots. If it’s not blocking, nothing will be installed (just downloaded) and the system rebooted all the same.

In sum, as I see it, **making not calling a GUI the default (either as an adjustment period or ever) is preferable to making GUI installs non-blocking**.

**Comment by jawshooah** [comment]
> In sum, as I see it, making not calling a GUI the default (either as an adjustment period or ever) is preferable to making GUI installs non-blocking.

In that case, I'd prefer that we make this an opt-in feature. Many of our users use scripts to automate the installation of many casks at once, usually to bootstrap a new machine. If we change the default behavior to opening GUI installers and blocking on user input, we're violating those users' expectations in a big way, and probably setting ourselves up for a nice influx of "bug" reports.

**Comment by vitorgalvao** [comment]
Changing the suggestion, then, and updating the top post. Making non-GUI the default, and add a `--yes-gui`. Changing that behaviour to be the reverse is a suggestion for another time (if ever).

**Comment by vitorgalvao** [comment]
Making `--yes-gui` `--allow-gui` instead (less ridiculous).

**Comment by vitorgalvao** [comment]
Change from `discussion` to `ready to implement`.

**Comment by yurikoles** [comment]
Does anyone volunteer mentorship on this ticket? I have small xp in Ruby, I'm an iOS Dev by myself with 6 years of xp.

**Comment by reitermarkus** [comment]
@yurikoles, if you decide to tackle this, we will certainly provide some guidance.

**Comment by yurikoles** [comment]
@reitermarkus what is the better way to communicate?

**Comment by yurikoles** [comment]
@reitermarkus I had dropped you an email.

**Comment by aaronliu0130** [comment]
I think we should also add an env var that'll basically automatically set the `--allow-gui` flag.

**Comment by p-linnane** [comment]
Closing this out due to age. If anyone wants to sort this out, a PR in Homebrew/brew would be welcome.
